eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentEvents.scala

```diff
@@ -112,10 +112,12 @@ object PaymentReceived {
 
 case class PaymentSettlingOnChain(id: UUID, amount: MilliSatoshi, paymentHash: ByteVector32, timestamp: Long = System.currentTimeMillis) extends PaymentEvent
 
-sealed trait PaymentFailure
+sealed trait PaymentFailure {
+  def route: Seq[Hop]
+}
 
 /** A failure happened locally, preventing the payment from being sent (e.g. no route found). */
-case class LocalFailure(t: Throwable) extends PaymentFailure
+case class LocalFailure(route: Seq[Hop], t: Throwable) extends PaymentFailure
```

Comment:
I tried it in https://github.com/ACINQ/eclair/pull/1414/commits/c4e7ed773ea6f3e7f8090ee5d127d4a81e9ed2da but I'm not entirely convinced, let me know if I should keep the commit.
The issue is that it still needs the developer to make sure it uses the right type (retriable vs non-retriable). I think if we want to go in that direction we should go deeper and have a `Retriable` trait that is shared across the codebase. The issue is that some `ChannelExceptions` as well as most `RouterExceptions` need to extend that trait, and we can't easily rely on `Status.Failure(t: Exception)` in `PaymentLifecycle`, we would need to have a better typing of that exception which is a bit of a rabbit hole :/

Reply:
I tried the in-depth approach, and it feels a bit riskier and will deserve more testing. It would be a good clean-up of our usage of `Status.Failure` and arbitrary exceptions in some places of the codebase, but I think it deserves its own, separate PR. I'm leaning towards removing that commit for now and doing this work in a separate PR (probably after the MPP split v2).

Reply:
Maybe it's overkill, I just don't like using `Nil` for when there is no route. The model allows building a `LocalFailure(a :: b :: Nil, RouteNotFound)` which should not exist.

Reply:
The issue is that "no route" isn't equivalent to "non-retriable", because some local failures are non-retriable but do have a route (for example when an HTLC times out on-chain).
I can separate local failures that have a route from those that don't, but I'm afraid what we gain in avoiding a potential incoherency is lost in the additional pattern matching efforts elsewhere...

Reply:
Fair enough

Reply:
I have an action item to re-work the error interfaces between payment, router and channel though, just not for this PR :)